lib/rust/parser/src/syntax/tree.rs

```diff
@@ -194,11 +230,140 @@ impl<'s> span::Builder<'s> for Error {
 }
 
 
+// == Code Blocks ===
+
+/// A line of code.
+#[derive(Debug, Clone, PartialEq, Eq, Visitor, Reflect, Serialize, Deserialize)]
+pub struct Line<'s> {
+    /// The content of the line, if any.
+    pub expression: Option<Tree<'s>>,
+    /// The end-of-line token.
+    pub newline:    token::Newline<'s>,
+}
+
+impl<'s> Line<'s> {
+    /// Transform the content of the line with the provided function, if any is present; return the
+    /// result.
+    pub fn map_expression(self, f: impl FnOnce(Tree<'s>) -> Tree<'s>) -> Self {
+        let Self { newline, expression } = self;
+        let expression = expression.map(f);
+        Self { newline, expression }
+    }
+}
+
+impl<'s> From<token::Newline<'s>> for Line<'s> {
+    fn from(newline: token::Newline<'s>) -> Self {
+        Self { expression: None, newline }
+    }
+}
+
+impl<'s> span::Builder<'s> for Line<'s> {
+    fn add_to_span(&mut self, span: Span<'s>) -> Span<'s> {
+        span.add(&mut self.newline).add(&mut self.expression)
```

Comment:
wait, so the newline is first? It is not in the type definition.

Reply:
The inconsistency here was a mistake--I'm surprised to find the spans work out the same if these are in the wrong order, but looking at the implementation of `Span::concat_mut`, it happens to be commutative in some common cases.